cylc/flow/exceptions.py

```diff
@@ -397,30 +397,34 @@ class PlatformLookupError(CylcConfigError):
 
 
 class HostSelectException(CylcError):
-    """No hosts could be selected from the provided configuration."""
+    """No hosts could be selected from the provided configuration.
 
-    def __init__(self, data: Dict[str, dict]):
+    Args:
+        data: Mapping of hostnames to error info.
+        ranking: The ranking expression used to select the host.
+    """
+
+    def __init__(self, data: Dict[str, dict], ranking: Optional[str] = None):
         self.data = data
+        self.ranking = ranking
 
     def __str__(self) -> str:
         ret = 'Could not select host from:'
         for host, data in sorted(self.data.items()):
-            if host != 'ranking':
```

Comment:
You've removed a bunch of guards against "ranking".

Do you understand what these were intended to do?

Reply:
I found it confusing that ranking information was included in a mapping of hostnames to exception messages. It's been pulled out into a separate field on `HostSelectException`.